File: Hackerearth/Dynamic Programming/Graph/Graph Representation/edge_existence.cpp
/*
                C++
                encoding:UTF-8
                Modified: <31/Dec/2018 10:40:54 PM>

                ✪ H4WK3yE乡
                Mohd. Farhan Tahir
                Indian Institute Of Information Technology (IIIT),Gwalior
*/

#include <bits/stdc++.h>
using namespace std;
#define ll long long

int main() {
  ios_base::sync_with_stdio(false), cin.tie(0), cout.tie(0);
  ll n, m;
  cin >> n >> m;
  std::vector<ll> graph[n + 1];
  ll a, b;
  for (size_t i = 0; i < m; i++) {
    cin >> a >> b;
    graph[a].push_back(b);
    graph[b].push_back(a);
  }
  ll q;
  cin >> q;
  for (size_t i = 0; i < q; i++) {
    cin >> a >> b;
    int flag = 0;
    for (size_t j = 0; j < graph[a].size(); j++) {
      if (graph[i][j] == b) {
        std::cout << "YES" << '\n';
        flag = 1;
        break;
      }
      if (flag == 0)
        std::cout << "NO" << '\n';
    }
  }
  return 0;
}

File: Hackerearth/Dynamic Programming/Graph/Graph Representation/edge_existence.py
#	!/bin/env python3
#	encoding: UTF-8
#	Modified: <31/Dec/2018 10:47:06 PM>


#	✪ H4WK3yE乡
#	Mohd. Farhan Tahir
#	Indian Institute Of Information Technology (IIIT),Gwalior

#	Question Link
#
#

# ///==========Libraries, Constants and Functions=============///


import sys

inf = float("inf")
mod = 1000000007


def get_array(): return list(map(int, sys.stdin.readline().split()))


def get_ints(): return map(int, sys.stdin.readline().split())


def input(): return sys.stdin.readline()

# ///==========MAIN=============///


def main():
    from collections import defaultdict
    Graph = defaultdict(list)
    n, m = get_ints()
    for i in range(m):
        a, b = get_ints()
        Graph[a].append(b)
        Graph[b].append(a)
    for q in range(int(input())):
        a, b = get_ints()
        if b in Graph[a]:
            print('YES')
        else:
            print('NO')


if __name__ == "__main__":
    main()
